src/lib/trace.c

```diff
@@ -69,494 +70,110 @@ static void put_header(volatile uint32_t *dst, uint64_t timestamp)
-	dt[payload_offset + 2] = param2;
-	dt[payload_offset + 3] = param3;
-	dtrace_event((const char *)dt, sizeof(uint32_t) * message_size_dwords);
-
-	/* send event by mail box too. */
-	spin_lock_irq(&trace->lock, flags);
-
-	/* write timestamp and event to trace buffer */
-	t = (volatile uint32_t *)(MAILBOX_TRACE_BASE + trace->pos);
-	trace->pos += sizeof(uint32_t) * message_size_dwords;
-
-	if (trace->pos > MAILBOX_TRACE_SIZE
-		- sizeof(uint32_t) * message_size_dwords)
-		trace->pos = 0;
-
-	spin_unlock_irq(&trace->lock, flags);
-
-	put_header(t, time);
-	t[payload_offset] = log_entry;
-	t[payload_offset + 1] = param1;
-	t[payload_offset + 2] = param2;
-	t[payload_offset + 3] = param3;
-
-	/* writeback trace data */
-	dcache_writeback_region((void *)t,
-		sizeof(uint32_t) * message_size_dwords);
-}
-
-void _trace_event_mbox_atomic3(uint32_t log_entry, uint32_t param1,
-	uint32_t param2, uint32_t param3)
-{
-	volatile uint32_t *t;
-	uint32_t message_size_dwords = MESSAGE_SIZE(3);
-	uint32_t payload_offset = sizeof(struct log_entry_header)
-		/ sizeof(uint32_t);
-	uint32_t dt[MESSAGE_SIZE(3)];
-	uint64_t time;
-
-	if (!trace->enable)
-		return;
-
-	time = platform_timer_get(platform_timer);
-
-	put_header(dt, time);
-	dt[payload_offset] = log_entry;
-	dt[payload_offset + 1] = param1;
-	dt[payload_offset + 2] = param2;
-	dt[payload_offset + 3] = param3;
-	dtrace_event_atomic((const char *)dt,
-		sizeof(uint32_t) * message_size_dwords);
-
-	/* write timestamp and event to trace buffer */
-	t = (volatile uint32_t *)(MAILBOX_TRACE_BASE + trace->pos);
-	trace->pos += sizeof(uint32_t) * message_size_dwords;
-
-	if (trace->pos > MAILBOX_TRACE_SIZE
-		- sizeof(uint32_t) * message_size_dwords)
-		trace->pos = 0;
-
-	put_header(t, time);
-	t[payload_offset] = log_entry;
-	t[payload_offset + 1] = param1;
-	t[payload_offset + 2] = param2;
-	t[payload_offset + 3] = param3;
-
-	/* writeback trace data */
-	dcache_writeback_region((void *)t,
-		sizeof(uint32_t) * message_size_dwords);
-}
+// TODO: it should be possible to for arbitrary array, passed as argument
+#define _TRACE_EVENT_NTH_IMPL_DT_STEP(i, _) dt[payload_offset + 1 + i] = CONCAT(param, i);
+#define _TRACE_EVENT_NTH_IMPL_T_STEP(i, _)   t[payload_offset + 1 + i] = CONCAT(param, i);
+#define _TRACE_EVENT_NTH_IMPL_DT(arg_count) SEQ_FROM_0_TO(arg_count, _TRACE_EVENT_NTH_IMPL_DT_STEP)
+#define _TRACE_EVENT_NTH_IMPL_T(arg_count)  SEQ_FROM_0_TO(arg_count, _TRACE_EVENT_NTH_IMPL_T_STEP)
+
+/* _trace_event function with poor people's version of constexpr if */
+#define _TRACE_EVENT_NTH_IMPL(is_mbox, is_atomic, arg_count)\
+_TRACE_EVENT_NTH(CONCAT(\
+IF_ELSE(is_mbox)(_mbox)(),   /*if is_mbox  , append _mbox   to function name*/\
+IF_ELSE(is_atomic)(_atomic)()/*if is_atomic, append _atomic to function name*/\
+), arg_count)\
+{\
+	uint32_t MESSAGE_SIZE_dwords = MESSAGE_SIZE(arg_count);\
+	uint32_t payload_offset = sizeof(struct log_entry_header)\
+		/ sizeof(uint32_t);\
+	uint32_t dt[MESSAGE_SIZE_dwords];\
```

Comment:
This values should come directly from MESSAGE_SIZE() macro expansion, to make it compile-time constant. Otherwise compiler wouldn't be able to optimize this, and we'll end up with same issue that we had recently. Please refer to current implementation.

Reply:
From what I see it should be compiler-optimizable, but you're right that by changing it to
`uint32_t dt[MESSAGE_SIZE(arg_count)]`
we are making one "should optimize" into "will optimize".

Reply:
We've analyzed the assembly with code that "should optimize" and even with -O3 -Os it didn't work as intended. Please refer to https://github.com/thesofproject/sof/pull/435 which fixed the issue.